File: templates/code-carl/ch05_stack/t1_lc0232_my_queue.py
"""LeetCode 232. Implement Queue using Stacks
https://leetcode.cn/problems/implement-queue-using-stacks/
"""


class MyQueue:
    def __init__(self) -> None:
        pass

    def push(self, x: int) -> None:
        pass

    def pop(self) -> int:
        pass

    def peek(self) -> int:
        pass

    def empty(self) -> bool:
        pass


if __name__ == "__main__":
    queue = MyQueue()
    queue.push(1)
    queue.push(2)
    assert queue.peek() == 1
    assert queue.pop() == 1
    assert queue.empty() is False
    assert queue.pop() == 2
    assert queue.empty() is True
    print("All tests passed.")

File: solutions/java/code-carl/ch05_stack/T1_LC0232_MyQueue.java
/**
 * Practice template for:
 * - LeetCode 232. Implement Queue using Stacks
 *   https://leetcode.cn/problems/implement-queue-using-stacks/
 */
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

class T1_LC0232_MyQueue {
    Deque<Integer> stackIn;
    Deque<Integer> stackOut;

    public T1_LC0232_MyQueue() {
        this.stackIn = new ArrayDeque<>();
        this.stackOut = new ArrayDeque<>();
    }

    public void push(int x) {
        stackIn.push(x);
    }

    public int pop() {
        if (stackOut.isEmpty()) {
            while (!stackIn.isEmpty()) {
                stackOut.push(stackIn.pop());
            }
        }
        return stackOut.pop();
    }

    public int peek() {
        int pop = this.pop();
        stackOut.push(pop);
        return pop;
    }

    public boolean empty() {
        return stackIn.isEmpty() && stackOut.isEmpty();
    }
}
